src/Nethermind/Nethermind.BlockValidation/Handlers/ValidateBuilderSubmissionHandler.cs

```diff
@@ -0,0 +1,331 @@
+        int totalBlobsLength = transactions.Sum(t => t.BlobVersionedHashes!.Length);
+
+        if (totalBlobsLength != blobsBundle.Blobs.Length)
+        {
+            error = $"Total blobs length mismatch. Expected {totalBlobsLength} but got {blobsBundle.Blobs.Length}";
+            return false;
+        }
+
+        if (totalBlobsLength != blobsBundle.Commitments.Length)
+        {
+            error = $"Total commitments length mismatch. Expected {totalBlobsLength} but got {blobsBundle.Commitments.Length}";
+            return false;
+        }
+
+        if (totalBlobsLength != blobsBundle.Proofs.Length)
+        {
+            error = $"Total proofs length mismatch. Expected {totalBlobsLength} but got {blobsBundle.Proofs.Length}";
+            return false;
+        }
+
+        if (!KzgPolynomialCommitments.AreProofsValid(blobsBundle.Proofs, blobsBundle.Commitments, blobsBundle.Blobs))
+        {
+            error = "Invalid KZG proofs";
+            return false;
+        }
+
+        error = null;
+
+        _logger.Info($"Validated blobs bundle with {totalBlobsLength} blobs, commitments: {blobsBundle.Commitments.Length}, proofs: {blobsBundle.Proofs.Length}");
+
+        return true;
+    }
+
+    private BlockProcessor CreateBlockProcessor(IWorldState stateProvider)
+    {
+        return new BlockProcessor(
+            _txProcessingEnv.SpecProvider,
+            _blockValidator,
+            new Consensus.Rewards.RewardCalculator(_txProcessingEnv.SpecProvider),
+            new BlockProcessor.BlockValidationTransactionsExecutor(_txProcessingEnv.TransactionProcessor, stateProvider),
+            stateProvider,
+            _receiptStorage,
+            new BlockhashStore(_txProcessingEnv.SpecProvider, stateProvider),
+            _txProcessingEnv.LogManager,
+            new WithdrawalProcessor(stateProvider, _txProcessingEnv.LogManager!),
+            new ReceiptsRootCalculator()
+        );
+    }
+
+    private static void FinalizeStateAndBlock(IWorldState stateProvider, Block processedBlock, IReleaseSpec currentSpec, Block currentBlock, IBlockTree blockTree)
+    {
+        stateProvider.StateRoot = processedBlock.StateRoot!;
+        stateProvider.Commit(currentSpec);
+        stateProvider.CommitTree(currentBlock.Number);
+        blockTree.SuggestBlock(processedBlock, BlockTreeSuggestOptions.ForceSetAsMain);
+        blockTree.UpdateHeadBlock(processedBlock.Hash!);
+    }
+
+
+    private bool ValidatePayload(Block block, Address feeRecipient, UInt256 expectedProfit, long registerGasLimit, bool useBalanceDiffProfit, bool excludeWithdrawals, out string? error)
+    {
+        if (!HeaderValidator.ValidateHash(block.Header))
+        {
+            error = $"Invalid block header hash {block.Header.Hash}";
+            return false;
+        }
+
+        if (!_blockTree.IsBetterThanHead(block.Header))
+        {
+            error = $"Block {block.Header.Hash} is not better than head";
+            return false;
+        }
```

Comment:
We can validate only on future blocks?

Reply:
Yeah it seems from this snippet: https://github.com/flashbots/builder/blob/df9c765067d57ab4b2d0ad39dbb156cbe4965778/core/blockchain.go#L2464

definition at: https://github.com/flashbots/builder/blob/df9c765067d57ab4b2d0ad39dbb156cbe4965778/core/forkchoice.go#L77